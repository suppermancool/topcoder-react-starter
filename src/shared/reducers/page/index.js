/**
 * Redux Reducer for state.page
 *
 * Description:
 *  Implements reducer factory for the state.page segment of Redux state; and
 *  combines it with the child state.page.x reducer factories.
 */

import { combineReducers } from 'redux';
import { redux } from 'topcoder-react-utils';
import sandbox from './sandbox';

/**
 * Reducer factory.
 * @param {Object} req ExpressJS HTTP Request.
 * @return {Function} Reducer.
 */
export function factory() {
  return redux.resolveReducers({
  }).then(reducers => combineReducers({
    ...reducers,
    sandbox,
  }));
}

export default combineReducers({
  sandbox,
});
