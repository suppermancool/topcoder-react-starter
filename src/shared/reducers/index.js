/**
 * This module holds the root factory for Redux reducer. The factory accepts
 * ExpressJS HTTP request object as its only optional argument, and it always
 * returns a promise that resolves to a new reducer.
 *
 * When the argument is provided, the factory assumes server-side rendering,
 * and it can use data available in the HTTP request to generate initial store
 * state fitted to the request.
 *
 * When no argument is provided, the factory assumes client-side rendering,
 * and creates the default state (in case of server-side rendering, that default
 * state will be overriden at client side by the one injected from the server).
 *
 * To better understand reducers read:
 * http://redux.js.org/docs/basics/Reducers.html.
 */

import _ from 'lodash';
import { redux } from 'topcoder-react-utils';
import { reducerFactory } from 'topcoder-react-lib';
import { factory as pageFactory } from './page';
import { factory as challengeListingFactory } from './challenge-listing';

export async function factory(req) {
  const resolvedReducers = await redux.resolveReducers({
    standard: reducerFactory(req),
    challengeListing: challengeListingFactory(req),
    page: pageFactory(req),
  });

  return redux.combineReducers((state) => {
    const res = { ...state };
    if (req) {
      res.domain = `${req.protocol}://${req.headers.host || req.hostname}`;
    }
    return res;
  }, {
    ..._.omit(resolvedReducers, 'standard'),
    ...resolvedReducers.standard,
  });
}

export default undefined;
