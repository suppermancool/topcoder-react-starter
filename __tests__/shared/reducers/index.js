import { factory } from 'reducers';

describe('Reducer created without HTTP request', () => {
  let reducer;

  beforeAll(async () => {
    reducer = await factory();
  });

  test('Creates expected initial state', () => {
    expect(reducer({}, {})).toMatchSnapshot();
  });
});


describe('Reducer created with HTTP request', async () => {
  let reducer;

  beforeAll(async () => {
    reducer = await factory({
      headers: {

      },
      hostname: 'test-website.com',
      protocol: 'http',
    });
  });

  test('Creates expected initial state', () => {
    expect(reducer).toMatchSnapshot();
  });
});
