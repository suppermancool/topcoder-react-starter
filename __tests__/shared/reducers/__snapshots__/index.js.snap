// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer created with HTTP request Creates expected initial state 1`] = `[Function]`;

exports[`Reducer created without HTTP request Creates expected initial state 1`] = `
Object {
  "auth": Object {
    "authenticating": true,
    "profile": null,
    "tokenV2": "",
    "tokenV3": "",
    "user": null,
  },
  "challenge": Object {
    "checkpoints": null,
    "details": null,
    "loadingCheckpoints": false,
    "loadingDetailsForChallengeId": "",
    "loadingResultsForChallengeId": "",
    "mySubmissions": Object {},
    "mySubmissionsManagement": Object {},
    "registering": false,
    "results": null,
    "resultsLoadedForChallengeId": "",
    "unregistering": false,
    "updatingChallengeUuid": "",
  },
  "challengeListing": Object {
    "allDraftChallengesLoaded": false,
    "allPastChallengesLoaded": false,
    "allReviewOpportunitiesLoaded": false,
    "challengeSubtracks": Array [],
    "challengeSubtracksMap": Object {},
    "challengeTags": Array [],
    "challenges": Array [],
    "expandedTags": Array [],
    "filter": Object {},
    "filterPanel": Object {
      "expanded": false,
      "searchText": "",
      "trackModalShown": false,
    },
    "keepPastPlaceholders": false,
    "lastRequestedPageOfDraftChallenges": -1,
    "lastRequestedPageOfPastChallenges": -1,
    "lastRequestedPageOfReviewOpportunities": -1,
    "lastUpdateOfActiveChallenges": 0,
    "loadingActiveChallengesUUID": "",
    "loadingChallengeSubtracks": false,
    "loadingChallengeTags": false,
    "loadingDraftChallengesUUID": "",
    "loadingPastChallengesUUID": "",
    "loadingReviewOpportunitiesUUID": "",
    "reviewOpportunities": Array [],
    "selectedCommunityId": "",
    "sidebar": Object {
      "activeBucket": "all",
      "activeSavedFilter": 0,
      "editSavedFiltersMode": false,
      "isSavingFilter": false,
      "savedFilters": Array [],
    },
    "sorts": Object {},
    "srms": Object {
      "data": Array [],
      "loadingUuid": "",
      "timestamp": 0,
    },
  },
  "direct": Object {
    "loadingProjectDetailsForId": 0,
    "loadingProjectsForUsername": "",
    "projectDetails": Object {},
    "projectPermissions": Object {
      "loadingForProjectId": 0,
      "permissions": Array [],
      "projectId": 0,
      "timestamp": 0,
    },
    "projects": Array [],
  },
  "errors": Object {
    "alerts": Array [],
    "icons": Object {
      "api": Array [],
      "network": Array [],
    },
  },
  "groups": Object {
    "groups": Object {},
    "loading": Object {},
  },
  "memberTasks": Object {
    "allLoaded": false,
    "lastRequestedPageNum": -1,
    "loadingUuid": "",
    "tasks": Array [],
    "timestamps": Object {},
  },
  "members": Object {},
  "mySubmissionsManagement": Object {},
  "page": Object {
    "sandbox": Object {
      "payments": Object {
        "editor": Object {
          "getMemberSuggestionsForKeyword": "",
          "memberInputKeyword": "",
          "memberInputPopupVisible": false,
          "memberInputSelected": Object {},
          "memberSuggestions": Array [],
          "pageState": "NEW_PAYMENT",
          "paymentAmount": 0,
          "paymentAssignee": "",
          "paymentDescription": "",
          "paymentTitle": "",
          "selectedBillingAccountId": 0,
          "selectedProjectId": 0,
        },
        "listing": Object {
          "selectedProjectId": 0,
        },
      },
    },
  },
  "profile": Object {
    "achievements": null,
    "copilot": false,
    "country": "",
    "info": null,
    "loadingError": false,
    "skills": null,
    "stats": null,
  },
  "reviewOpportunity": Object {
    "authError": false,
    "details": null,
    "isLoadingDetails": false,
    "requiredTerms": Array [],
  },
  "stats": Object {
    "communities": Object {},
  },
  "terms": Object {
    "selectedTerm": null,
    "terms": Array [],
  },
}
`;
